ls1intum/Artemis
# `General`: Open routing via chart in new tab

### Checklist
#### General


#### Client

### Motivation and Context


Currently, the access of other pages via charts is implemented in that way, that the page changes in the same browser tab. Since the click event is handled by the charting framework, we cannot implement the delegation as link without rebuilding the whole chart from scratch. Especially for charts that provide an overview of different aspects, this might be a suboptimal solution. The user would need to navigate back via browser navigation or breadcrumbs. Instead, I changed the navigation of selected charts in that way, that the new page opens in a separate tab. Then, the user can simply switch tabs to return to the overview.

### Description

I implemented a new routing type in `navigation.utils.ts` that handles the new routing behavior. I replaced the angular router with the auxiliary service within the components.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student
- 1 Exam with submissions
- 1 Programming exercise SCA enabled and submissions that got points deducted by SCA

1. Log in to Artemis as instructor
2. Navigate to the course statistics of the corresponding course
3. Select any bar in the average scores chart. The corresponding exercise statistics should open in a new tab
4. Navigate to the course scores. Selecting any bar should open a new tab with the participation scores view.
5. Navigate to the grading configuration page of the programming exercise. Select any bar part of the points chart in the test cases tab. The behavior should be similar
6. Switch to the SCA tab. Select any part of the lower two bars. Again, the new page should open in a new tab.
7. Navigate to the exam statistics. Select any bar in the score distribution. Make sure that a new tab opens
8. Select any of the horizontal bars in the charts below (only those concerning a specific exercise implement the routing). Here, a new tab should open again.
9. Log out and log in as Student again
10. Navigate to the course statistics tab. Check that the same behavior is observable for both the horizontal bar charts and the line chart.

### Review Progress





#### Code Review
#### Manual Tests

### Test Coverage


| Class/File | Branch | Line |
|------------|-------:|-----:|
| `course-scores.component.ts` | 73.88  | 94.72 |
| `exam-scores-average-score-graph.component.ts` | 70.21 | 84.9 |
| `exam-scores.component.ts` | 83.72 | 97.61 |
| `sca-category-distribution-chart.component.ts` | 90.32 | 88.4 |
| `test-case-distribution-chart.component.ts` | 93.33 | 88.46 |
| `exercise-assessment-dashboard.component.ts` | 66.95 | 81.32 |
| `course-statistics.component.ts` | 86.99 | 96.06 |
| `exercise-scores-chart.component.ts` | 79.41 | 98.05 |
| `statistics-average-score-graph.component.ts` | 74.41 | 93.38 |


### Screenshots


Small demo about the intended result:
https://user-images.githubusercontent.com/80622272/160232300-8cb324b6-73dd-4701-ae9d-f903997ed954.mp4
